### PR TITLE
Add support for initializing the value graph with the current graph

### PR DESCRIPTION
**Pull request by pepibumur**
### Short description 📝
Continuing with our effort of moving away from a reference-type graph, I'm extending the `ValueGraph` constructor to be initialized with the current graph. After we validate the format of the new graph with the current mappers, the mapping between the current and the new graph will be necessary as we update utilities to work with the `ValueGraph`:

```swift
let graph = ...
let valueGraph = ValueGraph(graph: graph)
generator.generate(graph: valueGraph)
```

### Solution 📦
- [x] Add constructor
- [x] Test constructor

### CODE REVIEW

File: Sources/TuistCore/ValueGraph/ValueGraph.swift [hunk 1]
```diff
@@ -16,15 +22,114 @@ public struct ValueGraph: Equatable {
...
+    ///   - node: Node whose dependencies will be mapped.
+    ///   - dependencies: Dictionary containing the dependency tree.
+    private static func map(node: GraphNode, into dependencies: inout [ValueGraphDependency: Set<ValueGraphDependency>]) {
+        let nodeDependency = dependency(from: node)
+        var nodeDependencies: Set<ValueGraphDependency>! = dependencies[nodeDependency]
+        if nodeDependencies == nil {
+            nodeDependencies = Set()
+        }
+
+        if let targetNode = node as? TargetNode {
```
Comment (kwridan) [inline Sources/TuistCore/ValueGraph/ValueGraph.swift:96]: minor suggestion:  You can leverage `switch` with type casting in many of those `if let as?`where we're dealing with multiple cases  _(not sure if its any better)_

e.g.

```swift
switch node {
   case let target as TargetNode:
       // ...
   case let frameworkNode as FrameworkNode:
       // ...
   case let xcframeworkNode as XCFrameworkNode:
      // ...
   default:
         break
 }
```

File: Tests/TuistCloudTests/ValueGraph/ValueGraphTests.swift [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import Foundation
```
Comment (kwridan) [inline Tests/TuistCloudTests/ValueGraph/ValueGraphTests.swift:1]: Should `ValueGraphTests.swift` be in `TuistCoreTests` ? Or were there going to be some `TuistCloud` specifics to test?

File: Sources/TuistCore/ValueGraph/ValueGraph.swift [hunk 3]
```diff
@@ -16,15 +22,114 @@ public struct ValueGraph: Equatable {
...
+                dependencies: [ValueGraphDependency: Set<ValueGraphDependency>]) {
+        self.name = name
+        self.path = path
         self.projects = projects
         self.packages = packages
         self.targets = targets
         self.dependencies = dependencies
     }
+
+    init(name: String, path: AbsolutePath, cache: GraphLoaderCaching) {
```
Comment (kwridan) [inline Sources/TuistCore/ValueGraph/ValueGraph.swift:41]: is there a need to also have a cache initialiser? May help remove some duplicate code if we limit it to `Graph` seeing we can make a graph from the cache.

Comment (pepibumur) [reply Sources/TuistCore/ValueGraph/ValueGraph.swift:41]: You are right, we can keep only the graph initializer.

File: Tests/TuistCloudTests/ValueGraph/ValueGraphTests.swift (hunk 2 above)
Comment (pepibumur) [reply Tests/TuistCloudTests/ValueGraph/ValueGraphTests.swift:1]: 😅 Yes... I'll move it

**Comment by pepibumur** [comment]
@kwridan I addressed your comments here and added a unit test to the constructor.

File: Sources/TuistCore/Graph/GraphLoaderCache.swift [hunk 4]
```diff
@@ -180,4 +183,40 @@ class GraphLoaderCache: GraphLoaderCaching {
     func targetNode(_ path: AbsolutePath, name: String) -> TargetNode? {
         targetNodes[path]?[name]
     }
+
+    func forEach(closure: (GraphNode) -> Void) {
```
Comment (kwridan) [inline Sources/TuistCore/Graph/GraphLoaderCache.swift:187]: minor: I believe this is no longer needed :)

**Pull request merged**